src/freenas/usr/local/sbin/migrate_warden.py

```diff
@@ -0,0 +1,400 @@
+                else:
+                    raise RuntimeError(
+                        f"ZFS pool '{self.pool}' is UNAVAIL!\n"
+                        f"Please check zpool status {self.pool} for more"
+                        " information.")
+
+        return True if match else False
+
+    def jail_exists(self):
+        """
+        :return: True if jail exists
+        """
+        try:
+            self.zfs.get_dataset(f"{self.pool}/iocage/jails/{self.jail}")
+        except libzfs.ZFSException as e:
+            if e.code == libzfs.Error.NOENT:
+                return False
+            else:
+                raise
+
+        return True
+
+    def send_dataset(self, send_fd, warden_dataset, date):
+        fromsnap = f"WardenMigration_{date}"
+
+        iocage_root = self.zfs.get_dataset(f"{self.pool}/iocage/jails/"
+                                           f"{self.jail}/root")
+        # We don't want this dataset anymore.
+        iocage_root.umount()
+        iocage_root.delete()
+
+        try:
+            warden_dataset.snapshot(
+                f"{warden_dataset.name}@WardenMigration_{date}",
+                recursive=True)
+        except libzfs.ZFSException as e:
+            if e.code == libzfs.Error.EXISTS:
+                pass  # Snapshot exists.
+            else:
+                raise
+
+        try:
+            warden_dataset.send(
+                send_fd,
+                fromname=None,
+                toname=fromsnap,
+                flags={
+                    None if not self.verbose else libzfs.SendFlag.PROGRESS,
+                    libzfs.SendFlag.PROPS
+                }
+            )
+        finally:
+            try:
+                os.close(send_fd)
+            except OSError:
+                pass
+
+    def recv_dataset(self, recv_fd, dataset):
+        # Defining these here instead of directly giving it in the function call
+        # as that aids in readability
+        force = True
+        nomnt = False
+        try:
+            self.zfs.receive(dataset, recv_fd, force, nomnt)
+        finally:
+            try:
+                os.close(recv_fd)
+            except OSError:
+                pass
+
+
+class Migrate(object):
+    def __init__(self, jail, _dir, pool, verbose, loop):
+        self.jail = jail
+        self.dir = _dir
+        self.meta = f"{self.dir}/.{self.jail}.meta"
+        self.pool = pool
+        self.dataset = f"{self.dir}/{self.jail}"
+        self.zfs = libzfs.ZFS(history=True,
+                              history_prefix="<warden-migration>")
+        self.ZFS = ZFS(self.pool, self.dataset, self.zfs, verbose)
+        self.thread_pool_executor = ThreadPoolExecutor(2)
+        self.loop = loop
+        self.r_pipe, self.s_pipe = os.pipe()
+        self.iocage_root = f"{self.pool}/iocage/jails/{self.jail}/root"
+        self.warden_dataset = self.zfs.get_dataset_by_path(self.dataset)
+        self.date = str(datetime.datetime.utcnow()).split()[0]
+
```

Comment:
I think we can skip the jailtype since we are only migrating jails and not old plugins to iocage. But the default routers / bridge ones should be accounted for as well.